cluster-autoscaler/proposals/resourceslack.md

```diff
@@ -0,0 +1,235 @@
+This document sketches how the resource slack feature will be implemented in a future release of K8S.
+
+## Motivation
+
+Currently, ClusterAutoscaler tries to maintain necessary and sufficient amount of nodes.
+For scaling up, it adds nodes only after some unschedulable pods are observed.
+This behavior of scaling up implies that, in a worst case, an unschedulable pod must wait until an eligible node is created, started, and then gets ready before it is finally scheduled.
+A node takes a much longer time than that of a pod to start and become ready and therefore it results in pods to occasionally delayed for several minutes to be horizontally scaled out.
+A possible work-around can be done today is to lower the threshold of HPA to trigger a scaling up of a replicaset/deployment earlier but it reduces total resource utilization when there are many replicasets and deployments required the work-around.
+Having the resource slack feature allows us to scale out pods faster while minimizing impact to resource utilization.
+
+## Design
+
+To achieve Resource Slack, introduce placeholder pods that will live only inside of ClusterAutoscaler.
+They are created to virtually request certain amounts of CPU and memory so that they influence all of the scale-up/scale-down decision made in CA, but won't be present in the scheduler running inside an actual K8S cluster, so the new pods will be able to pick the place of the placeholders and push the placeholders elsewhere (and make them trigger the scale-up/down).
+
+### Requirements
+
+Version 1
+
+* [R1] Works on relatively large clusters today's CA used to work (dozens of nodes, hundreds of pods? Correct me if necessary!)
+  * In general we don't want to have O(pods * nodes) operations every single loop
+* [R2] Keep the specified amount of extra capacity at minimum 
+* [R3] Ensure full compatibility with scale down
+* [R4] Only expose necessary and sufficient knobs to users
+
+Version 2
+
+* [R5] Don't add unnecessarily large node
+* [R6] Provide extra space even for the "biggest" pod in term of memory and cpu
+* [R7] Provide extra space even for every pod which is not biggest but do request some cpu(< max(pod_request_cpu)) and memory(< max(pod_request_mmory))
+  * Say, the "biggest pods" request 1000m cpu/1G mem, 10m cpu/10G mem respectively
+  * Even if they could fit within existing free space, a 500m(<1000m cpu)/5G mem pod has chances not to fit any node 
+
+Version 3
+
+* [R8] Reserve space even for the pods having tolerations
+* [R9] Reserve space even for the pods having nodeSelector/nodeAffinity
+ 
+### Specifications
+
+Version 1
+
+* [S1] On large cluster there is a huge number of pods to be scheduled. And the pods are mostly the same. If a pod from one replica set X failed to schedule on node A then there is no point to in trying to fit the next pod from X on A. Thus there should be just a single loop over all nodes (ref https://github.com/kubernetes/autoscaler/pull/56#discussion_r115942312) [R1]
```

Comment:
Sorry but I have no idea!
Is there any possible way to determine node A is out of capacity for a pod in replica set X?